### DISCUSSION TITLE
How do I implement the ability to detect rules for cs and go?

### DISCUSSION THREAD

**Discussion by RTxin** [pmd/pmd#5150] [Q&A] [answered]
If I want to add detection rules for `Go   or  C#`, do I need to construct files such as the ast directory from the `antlr.g4` file first, like swift?

According to the available information, this project already exists go & cs modules with `CSharpLexer.g4 & GoCpdLexer.jjava` and `Golang.g4 & CsCpdLexer.java` files.

I've had a general read of [adding_a_new_antlr_based_language.md](https://github.com/pmd/pmd/blob/master/docs/pages/pmd/devdocs/major_contributions/adding_a_new_antlr_based_language.md).

It's using Swift as an example, so if I want go/cs to be enabled do I also need to go through `Create AST node classes`, `Generate your parser (using ANTLR)`, etc?

But `public class CsLanguageModule extends CpdOnlyLanguageModuleBase`, which again seems like it should follow the documentation [adding_new_cpd_language.md](adding_new_cpd_language.md) To do this.

**Comment by jsotuyod** [comment, marked as answer]
Since there is already CPD support for the language (both C# and Go), there are a few steps you can skip, such as creating the parser or the submodule. 

Following the [guide to add a new PMD language](https://docs.pmd-code.org/latest/pmd_devdocs_major_adding_new_language_antlr.html):
* jump directly into step 3. The module (step 1) and the parser (step 2) are already in place.
* skip steps 4 and 5, as these are already taken care of, and jump into 6, creating the parser adapter.
* proceed to step 7 and 8
* step 9 is going to be slightly different. The `CsLanguageModule` already exists, but will need some changes. Specifically, extending `SimpleLanguageModuleBase` instead of `CpdOnlyLanguageModuleBase`.
  * The last item under step 9 (referencing the class from `META-INF/service/…` can be skipped, this is already done.
* All other remaining steps are still required

**Reply by RTxin** [reply]
Thank you for your answer.
